docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,322 @@
+point is occupied by prometheus-metrics-adapter.
+
+#### Heavyweight Prometheus infrastructure feeds signal back into cluster control loop
+The proposed signal flow brings two Prometheus instances (shoot prometheus and seed aggregate prometheus) into the shoot
+cluster control loop. This has major implications on reliability. The alternative is a small, custom-built component
+registered as seed custom metrics API extension, which directly calls the ShootKapi metrics endpoints and outputs
+request rate.
+
+The current approach was chosen solely because it leverages an existing open source component and avoids authoring a
+custom one. The historical record of Gardener Prometheus components reliability is deemed acceptable for the purpose.  
+
+The proposed complex signal path introduces additional latency to scaling response. At present, the resulting scaling
+latency is deemed satisfactory.
+
+While the scope of this GEP is limited to using the Prometheus signal path, the proposed design allows for an easy
+substitution with a signal path based on aforementioned simple custom component alternative. This supports future
+changes based on accumulated field experience, or evolving solution needs. 
+
+#### Observed fluctuations in actual compute efficiency affecting ShootKapi workloads
+There is a known source of horizontal scaling inefficiency, which presents an opportunity for substantial further
+improvement. It is not addressed by this proposal and instead deferred to a separate implementation stage, as it
+requires a strictly incremental effort on top of this GEP.
+
+An approximate 50% reduction (factor 0.5) of the effective compute power per unit of CPU was occasionally observed on
+some productive ShootKapi instances. The change occurs with a quick cutoff transition, and has been observed for periods
+or up to 2 hours. This variance currently forces substantial safety margin upon HPA's target metric value, effectively
+necessitating more replicas to be deployed than required during usual operation.
+
+The source is believed to be competition for compute resources with sibling workloads, either on the K8s, or
+infrastructure level, but the circumstances have not been conductive to investigation after the fact.
+A subsequent stage, outside the scope of this GEP is planned to focus on understanding the root cause, and further
+optimizing HPA's target metric value accordingly.
+
+Once this biggest source of inefficiency has been resolved, yet another round of scaling performance fine-tuning
+is planned, where instead of overall request rate, HPA will be driven by a weighted average, reflecting the different
+compute cost of different request categories (e.g. a cluster-scoped LIST is more cpu intensive than a resource-scoped PUT)
+
+## Alternatives
+Below is a list of the most promising alternatives to the proposed approach. 
+1. **Using VPA, plus a simple custom horizontal autoscaler which acts when VPA recommendation goes below or above a
```

Comment:
The sum of the memory usages might be a better metric than memory requests for this case.
When scaling up, the new replica will have the same memory request as the existing ones. Thus, it would immediately scale up again. However, the memory usage of the new instance should be less than the request.
For scaling down it is similar. 

While writing down these lines, I have the impression that this is pretty close to a scenario listed as limitation of VPA. 

>  Vertical Pod Autoscaler should not be used with the [Horizontal Pod Autoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#support-for-resource-metrics) (HPA) on CPU or memory at this moment. However, you can use VPA with [HPA on custom and external metrics](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#scaling-on-custom-metrics).

Reply:
We took this offline for a discussion over a broader communications channel. In the end, I guess it's valid to view the obstacle before driving HPA on memory requests/usage as a case of the officially declared HPA-VPA incompatibility, mentioned above by @oliver-goetz. We cannot control HPA/VPA hysteresis thresholds applied to the driving signal, so we cannot make them play nicely together.

Thanks to @istvanballok and @rickardsjp we had a chance for a broader discussion which touched on many aspects of autoscaling, and helped us realise the need to be outspoken in the GEP about the limitations imposed by the lack of proper multidimensional autoscaler.